src/lib/logical_query_plan/lqp_utils.cpp

```diff
@@ -506,4 +526,32 @@ void remove_invalid_fds(const std::shared_ptr<const AbstractLQPNode>& lqp, std::
    */
 }
 
+std::shared_ptr<AbstractLQPNode> find_diamond_origin_node(
+    const std::shared_ptr<AbstractLQPNode>& union_root_node) {
+  Assert(union_root_node->type == LQPNodeType::Union, "Expecting UnionNode as the diamond's root node.");
+  DebugAssert(union_root_node->input_count() > 1, "Diamond root node does not have two inputs.");
+  bool is_diamond = true;
+  std::optional<std::shared_ptr<AbstractLQPNode>> diamond_origin_node;
+  visit_lqp(union_root_node, [&](const auto& diamond_node) {
+    if (!is_diamond) return LQPVisitation::DoNotVisitInputs;
+    if (diamond_node->output_count() > 1) {
+      if (!diamond_origin_node.has_value()) {
```

Comment:
`if (!diamand_origin_node)`

Do we ever find structures like below, where the union node is the origin of another diamond, and thus, will be detected as the origin node when handed as `union_root_node`? Maybe we should only assign the origin node only when it is not the root. 
```
          Join
         /     \
Reduction       Anything
         \     /
          Union
         /     \
Predicate       Predicate
         \     /
         Anything
```

Reply:
Not sure, if I understand. We cannot call `find_diamond_origin_node` for other nodes than UnionNodes. When calling it for the Union of your LQP, it should return the bottom Anything node.

Reply:
Test added. Bug fixed :)